src/compute/src/server.rs

```diff
@@ -185,14 +243,149 @@ impl<'w, A: Allocate> Worker<'w, A> {
+        let idx = self.timely_worker.index();
+
+        {
+            let rx_activator = Arc::clone(&self.rx_activator);
+            let cmd_queue = Rc::clone(&cmd_queue);
+
+            self.timely_worker.dataflow::<u64, _, _>(move |scope| {
+                let src = source(scope, "Command source", |mut cap, info| {
+                    let activator = scope.sync_activator_for(&info.address[..]);
+                    {
+                        *rx_activator.lock().unwrap() = Some(activator);
+                    }
+
+                    move |output| {
+                        //println!("w{:?} Calling non-blocking recv", &idx);
+                        if idx == 0 {
+                            // We turn the TryRecvError into an Option type, as it doesnt
+                            // implemenet the right traits to fulfill Timelys ExchangeData
+                            //
+                            match command_rx.try_recv() {
+                                Ok(cmd) => {
+                                    // get some data and send it.
+                                    let time = cap.time().clone();
+                                    output.session(&cap).give(Some(cmd));
+
+                                    // downgrade capability.
+                                    cap.downgrade(&(time + 1));
+                                }
+                                Err(TryRecvError::Disconnected) => {
+                                    // get some data and send it.
+                                    let time = cap.time().clone();
+                                    output.session(&cap).give(None);
+
+                                    // downgrade capability.
+                                    cap.downgrade(&(time + 1));
+                                }
+                                Err(TryRecvError::Empty) => {
+                                    println!("w{:?} Active without data?", idx)
+                                }
+                            };
+                        }
+                    }
+                });
+
+                src.flat_map(move |x| {
+                    (0..peers)
+                        .zip(match x {
+                            Some(x) => Self::split_command(peers, x)
+                                .into_iter()
+                                .map(Some)
+                                .collect(),
+                            None => vec![None; peers],
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .exchange(|(dest, _)| *dest as u64)
+                .inspect(move |(_, cmd)| {
+                    println!("w{:?}, cmd={:?}", idx, cmd);
+                    (*cmd_queue).borrow_mut().push_back(match &*cmd {
+                        Some(x) => Ok(x.clone()),
+                        None => Err(TryRecvError::Disconnected),
+                    });
+                });
```

Comment:
Here, I'd recommend to write this in terms of `unary_frontier` so that the operator can supply its own exchange function to avoid the `exchange` operator, and that it gets to take ownership of the data it receives. It could reconstruct the `TryRecvError::Disconnected` from an empty frontier.